primitives/src/json.rs

```diff
@@ -0,0 +1,201 @@
+//! String containing serialized valid JSON.
+//! This string is guaranteed to parse as JSON
+
+#[cfg(not(feature = "std"))]
+use alloc::{
+    borrow::ToOwned,
+    string::{String, ToString},
+    vec::Vec,
+};
+use core::{
+    fmt::{Display, Formatter},
+    str::FromStr,
+};
+#[cfg(feature = "std")]
+use std::{
+    string::{String, ToString},
+    vec::Vec,
+};
+
+use iroha_schema::IntoSchema;
+use parity_scale_codec::{Decode, Encode};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use serde_json::Value;
+
+#[allow(missing_docs)]
+#[derive(Debug, Clone, Eq, Encode, Decode)]
+pub struct JsonString(pub String);
+
+/// A candidate for a valid `JsonString`.
+/// Is used for generalizing ser/de any types to `JsonString` and vise versa
+#[derive(Serialize, Deserialize, Clone)]
+struct JsonCandidate<T>(T);
+
+impl JsonString {
+    /// Construct [`JsonString`]
+    pub fn new<T: Serialize>(payload: T) -> Self {
+        JsonCandidate(payload).try_into().unwrap()
+    }
+
+    /// Tries cast [`JsonString`] to any value.
+    ///
+    /// # Errors
+    /// - if String == invalid Json type
+    pub fn try_into_any<T: DeserializeOwned>(&self) -> Result<T, serde_json::Error> {
+        serde_json::from_str(&self.0)
+    }
+}
+impl PartialEq for JsonString {
+    fn eq(&self, other: &Self) -> bool {
+        serde_json::from_str::<Value>(&self.0).unwrap()
+            == serde_json::from_str::<Value>(&other.0).unwrap()
+    }
+}
+
+impl<'de> serde::de::Deserialize<'de> for JsonString {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let json = Value::deserialize(deserializer)?;
+        Ok(Self(json.to_string()))
+    }
+}
+
+impl serde::ser::Serialize for JsonString {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let json: Value = serde_json::from_str(&self.0).map_err(serde::ser::Error::custom)?;
+        json.serialize(serializer)
+    }
+}
+
+impl iroha_schema::TypeId for JsonString {
+    fn id() -> iroha_schema::Ident {
+        "JsonString".to_owned()
+    }
+}
+
+impl IntoSchema for JsonString {
+    fn type_name() -> iroha_schema::Ident {
+        <Self as iroha_schema::TypeId>::id()
+    }
+
+    fn update_schema_map(map: &mut iroha_schema::MetaMap) {
+        if !map.contains_key::<Self>() {
+            map.insert::<Self>(iroha_schema::Metadata::String);
+        }
+    }
+}
+
+impl<T: Serialize> TryFrom<JsonCandidate<T>> for JsonString {
+    type Error = serde_json::Error;
+    fn try_from(value: JsonCandidate<T>) -> Result<Self, Self::Error> {
+        Ok(JsonString(serde_json::to_string(&value.0)?))
+    }
+}
+
+impl PartialOrd for JsonString {
+    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for JsonString {
```

Comment:
is this `Ord` implementation congruent with `PartialEq` implementation?

Reply:
@Stukalov-A-M it looks to me as if `Eq` and `Ord` are not incongruent

Reply:
> @Stukalov-A-M it looks to me as if `Eq` and `Ord` are not incongruent

Didn't catch the thought. Current implementation of `Ord` is fully consistent and realizes correct comparing for most of potential cases.
Other questions that regard to congruent we agreed to discuss in a separate issue https://github.com/hyperledger/iroha/pull/4732#discussion_r1642358856, isn't it?

Reply:
there is no such thing as correct comparing. Whatever you implement in `Ord` is "correct". What you have to make sure though is that when `Ord` returns `Ordering::Equal`, `Eq` must return true. Otherwise you break implementations of `BTreeSet` or `HashSet` etc. Since your `Ord` implementation looks specialized, I assume `Eq` should be specialized as well. You can read about which conditions must hold [here](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html)

Reply:
> there is no such thing as correct comparing. Whatever you implement in `Ord` is "correct"

How so?
```
let str1 = r#"{ "a" : 1.23}"#;
    let str2 = r#"{ "a" :      1.23}"#;

    println!("{:?}", str1.cmp(&str2));

    let str1 = serde_json::from_str::<Value>(str1).unwrap();
    let str2 = serde_json::from_str::<Value>(str2).unwrap();

    let value_1 = str1.to_string();
    let value_2 = str2.to_string();

    println!("{:?}", value_1.cmp(&value_2));

Greater
Equal

```

Reply:
@mversic , the congruity between `Ord`, `PartialOrd` and `PartialEq` is fixed.